// html-figma/browser/text-to-figma.js
import { fastClone, parseUnits, getRgb, defaultPlaceholderColor, } from '../utils';
import { getLineHeight, isHidden } from './dom-utils';
import { context } from './utils';
export const textToFigma = (node, { fromTextInput = false } = {}) => {
    var _a, _b, _c, _d, _e, _f, _g;
    const textValue = (_a = (node.textContent ||
        node.value ||
        node.placeholder)) === null || _a === void 0 ? void 0 : _a.trim();
    if (!textValue)
        return;
    const { getComputedStyle } = context.window;
    const parent = node.parentElement;
    if (isHidden(parent)) {
        return;
    }
    const computedStyles = getComputedStyle(fromTextInput ? node : parent);
    const range = context.document.createRange();
    range.selectNode(node);
    const rect = fastClone(range.getBoundingClientRect());
    const lineHeight = getLineHeight(node, computedStyles);
    range.detach();
    if (lineHeight && lineHeight.value && rect.height < lineHeight.value) {
        const delta = lineHeight.value - rect.height;
        rect.top -= delta / 2;
        rect.height = lineHeight.value;
    }
    if (rect.height < 1 || rect.width < 1) {
        return;
    }
    let x = Math.round(rect.left);
    let y = Math.round(rect.top);
    let width = Math.round(rect.width);
    let height = Math.round(rect.height);
    if (fromTextInput) {
        const borderLeftWidth = ((_b = parseUnits(computedStyles.borderLeftWidth)) === null || _b === void 0 ? void 0 : _b.value) || 0;
        const borderRightWidth = ((_c = parseUnits(computedStyles.borderRightWidth)) === null || _c === void 0 ? void 0 : _c.value) || 0;
        const paddingLeft = ((_d = parseUnits(computedStyles.paddingLeft)) === null || _d === void 0 ? void 0 : _d.value) || 0;
        const paddingRight = ((_e = parseUnits(computedStyles.paddingRight)) === null || _e === void 0 ? void 0 : _e.value) || 0;
        const paddingTop = ((_f = parseUnits(computedStyles.paddingTop)) === null || _f === void 0 ? void 0 : _f.value) || 0;
        const paddingBottom = ((_g = parseUnits(computedStyles.paddingBottom)) === null || _g === void 0 ? void 0 : _g.value) || 0;
        x = x + borderLeftWidth + (fromTextInput ? paddingLeft : 0);
        y = y + paddingTop;
        width = width - borderRightWidth - paddingRight;
        height = height - paddingTop - paddingBottom;
    }
    const textNode = {
        x,
        y,
        width,
        height,
        ref: node,
        type: 'TEXT',
        characters: (textValue === null || textValue === void 0 ? void 0 : textValue.replace(/\s+/g, ' ')) || '',
    };
    const fills = [];
    let rgb = getRgb(computedStyles.color);
    const isPlaceholder = fromTextInput &&
        !node.value &&
        node.placeholder;
    rgb = isPlaceholder ? defaultPlaceholderColor : rgb;
    if (rgb) {
        fills.push({
            type: 'SOLID',
            color: {
                r: rgb.r,
                g: rgb.g,
                b: rgb.b,
            },
            blendMode: 'NORMAL',
            visible: true,
            opacity: rgb.a || 1,
        });
    }
    if (fills.length) {
        textNode.fills = fills;
    }
    const letterSpacing = parseUnits(computedStyles.letterSpacing);
    if (letterSpacing) {
        textNode.letterSpacing = letterSpacing;
    }
    if (lineHeight) {
        textNode.lineHeight = lineHeight;
    }
    const { textTransform } = computedStyles;
    switch (textTransform) {
        case 'uppercase': {
            textNode.textCase = 'UPPER';
            break;
        }
        case 'lowercase': {
            textNode.textCase = 'LOWER';
            break;
        }
        case 'capitalize': {
            textNode.textCase = 'TITLE';
            break;
        }
    }
    const fontSize = parseUnits(computedStyles.fontSize);
    if (fontSize) {
        textNode.fontSize = Math.round(fontSize.value);
    }
    if (computedStyles.fontFamily) {
        // const font = computedStyles.fontFamily.split(/\s*,\s*/);
        textNode.fontFamily = computedStyles.fontFamily;
    }
    if (computedStyles.textDecoration) {
        if (computedStyles.textDecoration === 'underline' ||
            computedStyles.textDecoration === 'strikethrough') {
            textNode.textDecoration =
                computedStyles.textDecoration.toUpperCase();
        }
    }
    if (computedStyles.textAlign) {
        if (['left', 'center', 'right', 'justified'].includes(computedStyles.textAlign)) {
            textNode.textAlignHorizontal =
                computedStyles.textAlign.toUpperCase();
        }
    }
    return textNode;
};
